### PR TITLE
[jit] Add cache for MemoryDag::collectAllContainedMemoryLocations

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66295
* #65344
* #65178
* #65177
* __->__ #65122
* #65343

Failure to cache this seems to contribute to quadratic startup time for the static runtime.

Disclaimer: I am entirely un-versed in the performance considerations for the JIT and have no idea what the other impacts of this change may be. Let the reviewer beware.

Differential Revision: [D30983268](https://our.internmc.facebook.com/intern/diff/D30983268/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30983268/)!

### CODE REVIEW

**Comment by eellison** [comment]
will get to this tomorrow..

File: torch/csrc/jit/passes/utils/memory_dag.cpp [hunk 1]
```diff
@@ -189,6 +203,7 @@ void MemoryDAG::setWildcards(
   // For every element, if the cache contains `MemoryLocationFoo`, then we must
   // add `WildcardBar` to it.
   for (const std::unique_ptr<Element>& e : this->indexToElementMap_) {
+    e->cachedAllContainedMemoryLocations_.reset();
```
Comment (eellison) [inline torch/csrc/jit/passes/utils/memory_dag.cpp:206]: Why do we need to do this ?

File: c10/util/sparse_bitset.h [hunk 2]
```diff
@@ -548,6 +548,11 @@ class SparseBitVector {
     if (this == &RHS)
       return false;
 
+    if (empty()) {
+      *this = RHS;
```
[the remark this answers is not in the document]
Comment (swolchok) [reply c10/util/sparse_bitset.h:552]: it's an assignment operator call -- delegating to the assignment operator is a bit more efficient

File: torch/csrc/jit/passes/utils/memory_dag.cpp (hunk 1 above)
Comment (swolchok) [reply torch/csrc/jit/passes/utils/memory_dag.cpp:206]: we're mutating the graph, so we have to invalidate the caches (or update them, but that seems difficult to do efficiently)

**Comment by swolchok** [comment]
ping

**Comment by swolchok** [comment]
ping @eellison

**Review by eellison — COMMENTED** [review]
LGTM. The one sort of worry I had is it's interaction with our public mutating apis `replaceWithNewValue`, `copyValue`, `createValue`, but all of those apis retain the validity of the memory dag and strictly change the mapping from Value * -> Element *. I would still like to wait on @suo's review and it would also be helpful to add a test or two for the invalidation.

**Review by suo — APPROVED** [review]
This looks fine to me. lgtm as well. The cache invalidation is a touch conservative even—but probably not worth making more accurate.

**Pull request closed without merging**